### ISSUE TITLE
CURLE_SSL_CACERT (60) error when downloading tzdata2017c.tar.gz

### ISSUE THREAD

**Issue by karlosp** [HowardHinnant/date#260]
I get CURLE_SSL_CACERT (60) when downloading tzdata.

I made PR which fixed my problem https://github.com/HowardHinnant/date/pull/259

**Comment by HowardHinnant** [comment]
Merged: https://github.com/HowardHinnant/date/commit/5653e9e3a9806047e4be54cd1dab34de1c05e0d7

Thanks!

**Comment by crusader-mike** [comment]
Is it wise to switch off server certificate verification? It disables most of TSL connection security...

Also, you likely want to set `CURLOPT_NOSIGNAL` option too...

**Comment by HowardHinnant** [comment]
We're actually good here.  `curl_global_init` is only called by `curl_global()`, which is only used to initialize the function-local static `curl_is_now_initiailized` in `curl_init()`.  The function-local static ensures that only one thread can call this function, and then only once.

Of course my analysis assumes that this is the only library calling `curl_global_init`.  If other code outside of `date` is calling `curl_global_init`, that will be on the higher level code to ensure that one library or the other is initialized first.

**Comment by crusader-mike** [comment]
> Of course my analysis assumes that this is the only library calling curl_global_init.

Exactly my point -- problem is that your library might be used by another library and end-user code might be simply unaware of this. Basically libcurl poisoned your lib with this requirement and now you'll be poisoning any other code that will be using `date`. :)

I didn't spend enough time understanding code in full -- does this lib have "init" function or curl_global_init gets called at unspecified time on-demand?

**Comment by HowardHinnant** [comment]
The best way to preemptively "init" this lib is to just `date::get_tzdb()`.  One can even put that in a detached thread if that is convenient during start up:  `std::thread(date::get_tzdb).detach();`.  If other libs are trying to initialize curl at the same time, it would be best to create an application-wide initialize function that called `date::get_tzdb()` serially along with whatever other libs initialization triggers are.

**Comment by crusader-mike** [comment]
Does `date` documentation states that "call `date::get_tzdb()` to initialize this library before calling any other function, this call can't be made in parallel to any other code that could be calling `curl_global_init`"? ;-)

Now, any library that decides to use your library will have to introduce similar mechanism and propagate it up one layer for next user. You see what I mean? It is better to either fix damn libcurl or move responsibility of calling `curl_global_init` to upper layer (which means chance that he will forget and because `curl_easy_init` calls it under cover too -- you'll end up with mysterious impossible to reproduce crashes).

**Comment by crusader-mike** [comment]
btw... where is `curl_global_cleanup()` call?

**Comment by HowardHinnant** [comment]
I have no objection to documenting that this lib calls `curl_global_init` on first use.  I also have no objection to the authors of curl improving their thread safety behavior.  I don't see anything else I can do on my end.  If you see something else I can do, just let me know.

`curl_global_cleanup` is never called.  That would introduce further thread safety problems.  curl can be used at any time after initialization to do things such as checking the latest remote version of the IANA tzdb, or downloading a new version.  curl is intended to stay initialized for the life of the application.  The OS will reclaim all resources when the application ends.

**Comment by crusader-mike** [comment]
> I have no objection to documenting that this lib calls curl_global_init on first use

which means client code (that makes this first call) needs to make sure that `curl_global_init()` is not being called in parallel. Often it is plain impossible -- e.g. when your code is in library loaded dynamically (you just don't know what other host app threads could be doing right now). You'll end up with additional requirements like "after loading this lib you have to call this function on these conditions"... which may not be possible to meet for similar reasons.


> I don't see anything else I can do on my end.

Do not use libcurl, if you can. Do not call `curl_global_init()` -- explicitly make it user-code responsibility. This doesn't solve problem completely, but it will certainly help your client.


> The OS will reclaim all resources when the application ends.

Reminds me a case when some gamedev "defragmented" memory between loading levels -- he was restarting console and auto-started game with parameters that forced it to load next level. No, this is obviously wrong argument. For example, it will screw application that dynamically loads/unloads libraries,..

P.S. don't forget about CURL_NOSIGNAL -- that is another way in which libcurl is a gift that keeps giving.
P.S.S. Keep in mind that there are lot of curl versions out there that are full of bugs and crashes, especially ones compiled with libnss. I think latest version is fine (crossing fingers), but most installations in the wild don't use latest version.

**Comment by crusader-mike** [comment]
Oh, here is another one surprise (I just spent 2 days chasing it down): libcurl does not initialize OpenSSL -- it is user's responsibility. I don't see any `SSL_library_init()` calls in `date` library... It is a nicely hidden problem.

For info see [this](https://curl.haxx.se/libcurl/c/threadsafe.html). And libcurl can be built with many different "security providers" besides OpenSSL, for some of them proper initialization steps aren't even known.

In my case I have an app that uses AWS SDK (which uses libcurl) and second component (which uses libcurl too). You can't initialize libcurl twice without precautions -- initialization isn't thread-safe. But nice AWS SDK provides flags which say "I trust you've already initialized both libcurl and openssl" (which doesn't make a lot of sense when libcurl is built with libnsss or AWS SDK is built with WinSSL). Nevertheless, I found that there was an execution path that didn't "touch" second component and openssl was left uninitialized -- this resulted in very rare crashes. Real pain to track down.

And I am not done yet -- look at how openssl is meant to be [initialized](https://wiki.openssl.org/index.php/Library_Initialization) (depending on version). It's insane... It's initialization functions don't even return failures -- they always "succeed"! Now check how you uninitialize it (see Cleanup section) -- it just can't be done. I.e. you simply can't write a library that uses openssl and can be unloaded dynamically.

**Comment by rbock** [comment]
Hi Mike,

What is stopping you from downloading the tzdata in a separate executable and then work with the local version? As far as I can tell, that would avoid all the issues.

Then, what remains to be done: 

 1 Make the behavior clear in the documentation (what is initialized by the tz-library, what is not).
 1 Work with the vendors of libcurl and openssl (or whatever replacement you prefer) to improve the initialization calls on their end, i.e. attack the problem at the root, not at the leaf...

Just my 2ct.

Best,

Roland

PS: You mentioned earlier "Don't use libcurl [...]": What would be an easy to use alternative?

**Comment by crusader-mike** [comment]
> What is stopping you ...

Nothing. When I feel bored I entertain myself by looking for errors in projects I care about. Side effect is those projects may become slightly better and someone may avoid days of debugging. Lately I spent inordinate amount chasing bugs in libcurl/OpenSSL/libNSS and AWS SDK or in user code (due to poor decisions made by authors of aforementioned libs). It may made me cranky.

> You mentioned earlier "Don't use libcurl [...]": What would be an easy to use alternative?

I said:

> Do not use libcurl, if you can.

Unfortunately, there is no easy curl alternative that I know of. Beast library is dragging too many dependencies. In one of my projects I implemented my own subset of HTTP protocol. Naturally, it'll break the moment server on the other side changes it's behavior significantly. But on the bright side I spent only few days doing it -- significantly less than I spent chasing libcurl-related problems in another project.

**Comment by rbock** [comment]
Got it :-)

Well, as long as it cannot be changed in libcurl, openssl, etc, a more explicit documentation about side effects (like initializing curl, but not openssl or whatever other ssl library) seems to be the only real option without changing tz's behavior.

If changing tz's behavior is an option, then I would not init curl in the library, but leave that to the client code. Obviously, that would be a tough decision since it would likely break existing client code.

Btw: Why I am interested: I have a somewhat similar situation with underlying database connectors in sqlpp11. The mysql_library_init also has the trait of not being thread safe.

**Comment by crusader-mike** [comment]
> Well, as long as it cannot be changed in libcurl, openssl, etc, a more explicit documentation about side effects (like initializing curl, but not openssl or whatever other ssl library) seems to be the only real option without changing tz's behavior.

No, I think correct way to go is not to initialize anything that can't be safely initialized -- make it explicit requirement for end-user (since he is the only one who can safely do so). Also, note in docs (in big bold red letters) about using libraries like curl and associated problems (even though they are totally hidden from end user). Also note that user's libcurl may drag in similar "poisoners" like OpenSSL -- user code will have to solve non-trivial problem of how to detect which library libcurl is built with on a platform his code is currently being compiled with and then figure out how to correctly initialize it (based on version, phase of moon, etc).

> Btw: Why I am interested: I have a somewhat similar situation with underlying database connectors in sqlpp11. The mysql_library_init also has the trait of not being thread safe.

In my case I having in mind a small and thin "executor" that reads a "job" from file, loads related library (.so on Linux, .dll on Windows), executes certain exported function(s) and then after job is done -- unloads it. Tons of problems that shouldn't be there...

**Comment by HowardHinnant** [comment]
Closing as I don't see a clear action item for the time zone lib here.  I can not remove the initialization of things like curl and OpenSSL as it would break far more many users than it would fix.

**Comment by crusader-mike** [comment]
> I can not remove the initialization of things like curl and OpenSSL

Uhm...  you don't initialize OpenSSL -- that was one of issues raised in this thread. :)

**Comment by HowardHinnant** [comment]
Ok.  What's the action item for me?

**Comment by crusader-mike** [comment]
Here is the list of issues:
1. disabling certificate verification (via CURLOPT_SSL_VERIFYPEER), i.e. switching off most of protections TLS provides
    1a. on Windows (if libcurl is compiled with WinSSL instead if OpenSSL) you may also need to disable certificate revocation checks (as they won't work in closed environments), see CURLSSLOPT_NO_REVOKE
2. not using CURLOPT_NOSIGNAL, i.e. potentially breaking user code that uses certain signals
3. `curl_global_init` is not threadsafe, i.e. can break user code that initializes libcurl in other thread
4. libcurl does not initialize it's "security provider" library (OpenSSL, etc), i.e. potentially breaking user code
5. no cleanup for libcurl/OpenSSL/etc, i.e. potentially breaking user code that expects clean library (un)loading

Items 1-2 are trivial to address via code change. I see only two ways to properly fix items 3-5:
A) replace libcurl with smth else (maybe it is ok to download data using http? this will reduce scope by removing OpenSSL/etc out of picture)
B) or leave all (un)initialization (of libcurl/OpenSSL/etc) to user code -- i.e. do not attempt to init anything, clearly state this in user documentation

Note that (A) solves all 6 problems, but alas I don't know any good replacement. (B) would require breaking  changes to tz (which isn't too bad since it is already broken). If backward compatibility is an issue -- you can introduce v2 tz by either introducing another library or a v1/v2 flag that user can set to choose behavior.

Action item:
- having all this in mind figure out the best way to address issues 1-5 :-)
